### PR TITLE
Adding Blockly task problem

### PR DESCRIPTION
**Pull request by fthuin**
Hello,

Here are the main changes:

- `/inginious/common/tasks_code_boxes.py` BlocklyBox class added
- `/inginious/common/tasks_code_boxes.py` BlocklyProblem class added
- `/inginious/frontend/common/tasks_code_boxes.py` DisplayableBlocklyBox class added
- `/inginious/frontend/common/tasks_problems.py` DisplayableBlocklyProblem class added
- `/inginious/frontend/cmmon/templates/tasks/blockly.html` add a file with the blockly task HTML template
- `/inginious/frontend/common/static/js/` blockly javascript (see [Blockly repository](https://github.com/google/blockly)) + a javascript interpreter (see [JS-Interpreter](https://github.com/NeilFraser/JS-Interpreter) + some css and js to have a full screen modal
- `/inginious/frontend/webapp/templates/course_admin/edit_tabs/subproblem_templates.html` add the HTML to handle a blockly problem modification
- `/inginious/frontend/webapp/static/js/` javascript to handle the visual modification of tasks (toolbox and workspace of a problem)

Example of results:

selection_049

screencapture-localhost-8080-admin-pvz-edit-task-01_maze-1492676222586

selection_051

### CODE REVIEW

**Comment by GuillaumeDerval** [comment]
ok to test

File: inginious/common/tasks_code_boxes.py [hunk 1]
```diff
@@ -207,3 +207,32 @@ def __init__(self, problem, boxid, boxData):
...
+            self._workspace = boxData["workspace"]
+        if "options" in boxData:
+            self._options = boxData["options"]
+        if "files" in boxData:
+            self._files = boxData["files"]
+        if "blocks_files" in boxData:
+            self._blocks_files = boxData["blocks_files"]
+
+    def input_is_consistent(self, taskInput, default_allowed_extension, default_max_size):
+        if not MultilineBox.input_is_consistent(self, taskInput, default_allowed_extension, default_max_size):
```
Comment (GuillaumeDerval) [inline inginious/common/tasks_code_boxes.py]: Remove if; simply use `return MultilineBox.input_is_consistent(...)`

File: inginious/frontend/common/static/css/bootstrap-modal-fullscreen.css [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/* Copy-paste from http://www.minimit.com/demos/bootstrap-3-transparent-and-fullscreen-modals */
```
Comment (GuillaumeDerval) [inline inginious/frontend/common/static/css/bootstrap-modal-fullscreen.css]: Link not accessible. What about the license?

File: inginious/frontend/common/static/css/bootstrap-modal-transparent.css [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/* Copy-paste from http://www.minimit.com/demos/bootstrap-3-transparent-and-fullscreen-modals */
```
Comment (GuillaumeDerval) [inline inginious/frontend/common/static/css/bootstrap-modal-transparent.css]: Link not accessible; what about the license?

File: inginious/frontend/webapp/static/js/studio.js [hunk 4]
```diff
@@ -487,6 +493,296 @@ function studio_init_template_custom(well, pid, problem)
...
+ * @param problem
+ */
+function studio_init_template_blockly(well, pid, problem)
+{
+    var val = "";
+    if ("language" in problem) {
+        $('#language-' + pid, well).val(problem["language"]);
+    }
+    var toolbox_editor = registerCodeEditor($('#toolbox-' + pid)[0], 'xml', 10);
+    if ("toolbox" in problem) {
```
Comment (GuillaumeDerval) [inline inginious/frontend/webapp/static/js/studio.js]: This initialization code is far too looooong; consider using the ternary operator for giving default values:

```
toolbox_editor.setValue("toolbox" in problem ? problem["toolbox"] : '<xml xmlns="http://www.w3.org/1999/xhtml" style="display: none;"></xml>')
```

File: inginious/frontend/webapp/static/js/studio.js [hunk 5]
```diff
@@ -487,6 +493,296 @@ function studio_init_template_custom(well, pid, problem)
...
+    });
+
+}
+
+/**
+ * Create a new input for the name of a file
+ * @param pid
+ * @param filename
+ */
+function studio_create_javascript_files(pid, filename)
```
Comment (GuillaumeDerval) [inline inginious/frontend/webapp/static/js/studio.js]: Function name should reference blockly, as the function only serves this purpose

File: inginious/frontend/webapp/static/js/studio.js [hunk 6]
```diff
@@ -487,6 +493,296 @@ function studio_init_template_custom(well, pid, problem)
...
+        var script = $("<script>");
+        script.prop("charset", "utf-8");
+        script.prop("src", "/course/" + course + "/" + taskid + "/" + filename);
+        $("#collapse_" + pid).append(script);
+    } else {
+        $("#file-" + pid + "-" + index).val("");
+    }
+}
+
+function studio_create_blocks_files(pid, filename)
```
Comment (GuillaumeDerval) [inline inginious/frontend/webapp/static/js/studio.js]: Same here; add a reference to blockly in the function name.

File: inginious/frontend/webapp/templates/course_admin/edit_tabs/subproblem_templates.html [hunk 7]
```diff
@@ -284,6 +284,363 @@ <h4 style="text-align: center">Choices</h4>
...
+        <div id="blocklyFactory"></div>
+
+        <!-- Blockly official JavaScript -->
+        <script src="$get_homepath()/static/common/js/blockly/blockly_compressed.js" type="text/javascript" charset="utf-8"></script>
+        <script src="$get_homepath()/static/common/js/blockly/blocks_compressed.js" type="text/javascript" charset="utf-8"></script>
+        <script src="$get_homepath()/static/common/js/blockly/javascript_compressed.js" type="text/javascript" charset="utf-8"></script>
+        <script src="$get_homepath()/static/common/js/blockly/python_compressed.js" type="text/javascript" charset="utf-8"></script>
+
+        <!-- Blockly French Translation -->
+        <script src="$get_homepath()/static/common/js/blockly/msg/js/fr.js" type="text/javascript" charset="utf-8"></script>
```
Comment (GuillaumeDerval) [inline inginious/frontend/webapp/templates/course_admin/edit_tabs/subproblem_templates.html]: Either you do not include this or you provide a way to select the language.
For now INGInious is only in plain english; support for additionnal languages is under dev, see with @anthonygego.

File: inginious/frontend/common/static/css/bootstrap-modal-transparent.css (hunk 3 above)
Comment (fthuin) [inline inginious/frontend/common/static/css/bootstrap-modal-transparent.css]: This link is accessible for me. MIT License

File: inginious/frontend/common/static/css/bootstrap-modal-fullscreen.css (hunk 2 above)
Comment (fthuin) [inline inginious/frontend/common/static/css/bootstrap-modal-fullscreen.css]: This link is accessible for me. MIT License

Comment (GuillaumeDerval) [inline inginious/frontend/common/static/css/bootstrap-modal-fullscreen.css]: Mea culpa; it was the "HTTPS anywhere" plugin that forced the usage of https.

Where do you see it's MIT?

Comment (fthuin) [inline inginious/frontend/common/static/css/bootstrap-modal-fullscreen.css]: It is written in a lot of places on the website but I asked confirmation

selection_052

File: inginious/frontend/webapp/templates/course_admin/edit_tabs/subproblem_templates.html (hunk 7 above)
Comment (fthuin) [inline inginious/frontend/webapp/templates/course_admin/edit_tabs/subproblem_templates.html]: There is two ways I can handle it:
* Add an option in the problem to choose the language;
* Retrieve the language code from somewhere else (where?).

What do you think? @anthonygego

Comment (GuillaumeDerval) [inline inginious/frontend/webapp/templates/course_admin/edit_tabs/subproblem_templates.html]: Since the PR won't be merged anytime soon (we'll at least wait until June), let's wait until @anthonygego has merged the localisation system.

**Comment by anthonygego** [comment]
The code has been reworked. PR #266 and repo UCL-INGI/INGInious-blockly-plugin@a79ebcee030c7791d29eb7b3eb5f6bfa9c01cb91 are equivalent to this one.

Thank you very much for this contribution.
